frontend/src/js/entity-history/Timeline.tsx

```diff
@@ -419,14 +275,113 @@ export const Timeline = memo(
   },
 );
 
-const useTimeBucketedDataDesc = (
-  data: EntityHistoryStateT["currentEntityData"],
-): {
+const diffObjects = (objects: Object[]) => {
+  if (objects.length < 2) return {};
+
+  const differences: Record<string, Set<any>> = {};
+
+  for (let i = 0; i < objects.length - 1; i++) {
+    const o1 = objects[i] as any;
+    const o2 = objects[i + 1] as any;
+    const keys = Object.keys(o1);
+
+    for (const key of keys) {
+      if (
+        o1.hasOwnProperty(key) &&
+        o2.hasOwnProperty(key) &&
+        JSON.stringify(o1[key]) !== JSON.stringify(o2[key])
+      ) {
+        if (differences[key]) {
+          differences[key].add(o1[key]);
+          differences[key].add(o2[key]);
+        } else {
+          differences[key] = new Set([o1[key], o2[key]]);
+        }
+      }
+    }
+  }
+
+  return differences;
+};
+
+const findGroupsWithinQuarter =
+  (secondaryIds: ColumnDescription[]) =>
+  ({ quarter, events }: { quarter: string; events: EntityEvent[] }) => {
+    if (events.length < 2) {
+      return { quarter, groupedEvents: [events], differences: [{}] };
+    }
+
+    const groupedEvents: EntityEvent[][] = [[events[0]]];
+
+    for (let i = 1; i < events.length - 1; i++) {
+      const evt = events[i];
+      const lastEvt = groupedEvents[groupedEvents.length - 1][0];
+
+      const isDuplicateEvent = JSON.stringify(evt) === JSON.stringify(lastEvt);
+
+      if (isDuplicateEvent) {
+        continue;
+      }
+
+      const datesMatch =
+        JSON.stringify(evt.dates) === JSON.stringify(lastEvt.dates);
+      const sourcesMatch = evt.source === lastEvt.source;
+      const allSecondaryIdsMatch = secondaryIds.every(
+        ({ label }) => evt[label] === lastEvt[label],
+      );
+
+      const similarEvents = datesMatch && sourcesMatch && allSecondaryIdsMatch;
+
+      if (similarEvents) {
+        groupedEvents[groupedEvents.length - 1].push(evt);
```

Comment:
Since grouping is done by equality of a set of keys, you could use something like the following:

```ts
let groups  = new Map<[],EntityEvent[]>

for (event in events) {
  let key = [evt.source, event.dates].concat(secondaryIds.map(id => evt[id])]
  if (!(key in groups)) {
    groups[key] = []
  }
  groups[key].push(evt)
}
```

Reply:
Hmm I'm sorry I don't understand what you're saying

> Since grouping is done by equality of a set of keys
grouping is done by equality of a set of values, not keys here

what problem are you trying to address?